### PR TITLE
Batch processor enhancemenst through raw data parameter

### PR DESCRIPTION
**Pull request by RafalSkolasinski**
**What this PR does / why we need it**:

These solves family of issues gathered together under https://github.com/SeldonIO/seldon-core/issues/3702

**Which issue(s) this PR fixes**:

Closes https://github.com/SeldonIO/seldon-core/issues/3702

and therefore as well

Closes https://github.com/SeldonIO/seldon-core/issues/2657
Closes https://github.com/SeldonIO/seldon-core/issues/3409
Closes https://github.com/SeldonIO/seldon-core/issues/3681
Closes https://github.com/SeldonIO/seldon-core/issues/3408

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
```

### CODE REVIEW

**Comment by seldondev** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **rafalskolasinski**
You can assign the PR to them by writing `/assign @rafalskolasinski` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://jenkins-x.io/v3/develop/reference/chatops/?repo=SeldonIO%2Fseldon-core).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/SeldonIO/seldon-core/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by RafalSkolasinski** [comment]
Example on echo model for input:
```
{"data": {"names": ["a", "b", "c"], "ndarray": [[1, 2, 3]]}, "meta": {"tags": {"customer-id": 0}}}
{"data": {"names": ["a", "b", "c"], "ndarray": [[4, 5, 6]]}, "meta": {"tags": {"customer-id": 1}}}
{"data": {"names": ["a", "b", "c"], "ndarray": [[7, 8, 9]]}, "meta": {"tags": {"customer-id": 2}}}
{"data": {"names": ["a", "b", "c"], "ndarray": [[1, 3, 6]]}, "meta": {"tags": {"customer-id": 3}}}
```

run as
```
export SELDON_BATCH_WORKERS=100
export SELDON_BATCH_RETRIES=3

export SELDON_BATCH_HOST="localhost:9000"
export SELDON_BATCH_GATEWAY_TYPE=seldon

export SELDON_BATCH_DEPLOYMENT_NAME=iris
export SELDON_BATCH_NAMESPACE=seldon

export SELDON_BATCH_METHOD=predict
export SELDON_BATCH_TRANSPORT=rest

export SELDON_BATCH_DATA_TYPE=raw

export SELDON_BATCH_INPUT_DATA_PATH=input-new.txt
export SELDON_BATCH_OUTPUT_DATA_PATH=predict-output-new.txt

export SELDON_BATCH_SIZE=1

seldon-batch-processor
```


will give as output
```
{"data": {"names": ["t:0", "t:1", "t:2"], "ndarray": [[1, 2, 3]]}, "meta": {"tags": {"batch_id": "3eea7a8c-3bce-11ec-93bd-f799141f921b", "extra-tag": 10, "customer-id": 0, "batch_index": 0, "batch_instance_id": "3eeaad5e-3bce-11ec-93bd-f799141f921b-item-0"}}}
{"data": {"names": ["t:0", "t:1", "t:2"], "ndarray": [[4, 5, 6]]}, "meta": {"tags": {"batch_id": "3eea7a8c-3bce-11ec-93bd-f799141f921b", "extra-tag": 10, "customer-id": 1, "batch_index": 1, "batch_instance_id": "3eeaad5e-3bce-11ec-93bd-f799141f921b-item-1"}}}
{"data": {"names": ["t:0", "t:1", "t:2"], "ndarray": [[7, 8, 9]]}, "meta": {"tags": {"batch_id": "3eea7a8c-3bce-11ec-93bd-f799141f921b", "extra-tag": 942, "customer-id": 2, "batch_index": 2, "batch_instance_id": "3eeaad60-3bce-11ec-93bd-f799141f921b-item-0"}}}
{"data": {"names": ["t:0", "t:1", "t:2"], "ndarray": [[1, 3, 6]]}, "meta": {"tags": {"batch_id": "3eea7a8c-3bce-11ec-93bd-f799141f921b", "extra-tag": 942, "customer-id": 3, "batch_index": 3, "batch_instance_id": "3eeaad60-3bce-11ec-93bd-f799141f921b-item-1"}}}
```

(the `extra-tag` is randomly assigned from model)

File: python/seldon_core/batch_processor.py
Comment (axsaucedo) [inline python/seldon_core/batch_processor.py:338]: May be worth just making a note that we ignore anything inside meta outside tags, I think this makes sense, but just in case someone expects (for some strange reason) other fields like metrics not being carried out

File: python/seldon_core/batch_processor.py
Comment (axsaucedo) [inline python/seldon_core/batch_processor.py:351]: Is this all_equal function impl above a more efficient way, or why not just a simple list comprehension?

File: python/seldon_core/batch_processor.py
Comment (axsaucedo) [inline python/seldon_core/batch_processor.py:353]: else doesn't seem needed

File: python/seldon_core/batch_processor.py
Comment (axsaucedo) [inline python/seldon_core/batch_processor.py:358]: Nice one - we should also document this explicitly

File: python/seldon_core/batch_processor.py
Comment (axsaucedo) [inline python/seldon_core/batch_processor.py:335]: Feels like all this try block could be extracted into separate function as it's quite extensive

File: python/seldon_core/batch_processor.py
Comment (axsaucedo) [inline python/seldon_core/batch_processor.py:412]: Would batch id not be part of this meta?

**Review by axsaucedo — CHANGES_REQUESTED** [review]
Nice one - looks good overall, added couple of comments, would be good to also add documenattion as part of the PR in the batching section to cover the constraints in which this would work as weell

File: python/seldon_core/batch_processor.py
Comment (RafalSkolasinski) [reply python/seldon_core/batch_processor.py:338]: Docstring comment here?

File: python/seldon_core/batch_processor.py
Comment (RafalSkolasinski) [reply python/seldon_core/batch_processor.py:412]: Actually, it seems it would be only `batch_id`
```
    tags = {
        "batch_id": batch_id,
    }
```

This goes along what was happening in other `except Exception as e` block bellow

File: python/seldon_core/batch_processor.py
Comment (axsaucedo) [reply python/seldon_core/batch_processor.py:351]: Seems quite an obscure method - so why not just a simple list comprehsnion?

File: python/seldon_core/batch_processor.py
Comment (axsaucedo) [reply python/seldon_core/batch_processor.py:338]: primarily docs in batch page in seldon docs

File: python/seldon_core/batch_processor.py
Comment (RafalSkolasinski) [reply python/seldon_core/batch_processor.py:351]: Not sure how list comprehension would help? 
Alternative would be
```
names = names_list[0]
for other_name in names_list[1:]:
   assert other_name == names
```
(or similar)

File: python/seldon_core/batch_processor.py
Comment (yaliqin) [reply python/seldon_core/batch_processor.py:411]: Hi Rafal,  In my local code, before I found the error message is empty when there is 503 and 400 errors.  So I added something like the belows:
in the beginning of the method:
```
msg = ""
```,
in the requests part:
```
                seldon_payload = sc.predict(**predict_kwargs)
                msg = seldon_payload.msg
                logger.error(f"Exception: {e}, retries {i}, message:{msg}, request data:{data_np}")
```
At that time the request data "data_np" includes only data part or the "feature:value" depends on the input definition.   I find these information are quite useful for debugging then batch_id. Batch_id is generated  and hard to get the request input from that field if we look at the input file.

Comment (yaliqin) [reply python/seldon_core/batch_processor.py:411]: Hi Rafal,  In my local code, before I found the error message is empty when there is 503 and 400 errors.  So I added something like the belows:
in the beginning of the method:
```
msg = ""
```,
in the requests part:
```
                seldon_payload = sc.predict(**predict_kwargs)
                msg = seldon_payload.msg
                logger.error(f"Exception: {e}, retries {i}, message:{msg}, request data:{data_np}")
```
At that time the request data "data_np" includes only data part or the "feature:value" depends on the input definition.   I find these information are quite useful for debugging then batch_id. Batch_id is generated  and hard to get the request input from that field if we look at the input file.

File: python/seldon_core/batch_processor.py
Comment (RafalSkolasinski) [reply python/seldon_core/batch_processor.py:336]: done

File: python/seldon_core/batch_processor.py
Comment (RafalSkolasinski) [reply python/seldon_core/batch_processor.py]: done

File: python/seldon_core/batch_processor.py
Comment (RafalSkolasinski) [reply python/seldon_core/batch_processor.py:413]: We could print `batch_index` which basically corresponds to the line number

Comment (RafalSkolasinski) [reply python/seldon_core/batch_processor.py:413]: With that exact code snippet I am afraid that for
```
seldon_payload = sc.predict(**predict_kwargs)
msg = seldon_payload.msg
```
it may catch exception on first line and the `msg` printed may be not accurate

Comment (RafalSkolasinski) [reply python/seldon_core/batch_processor.py:413]: @yaliqin I added printing of batch indices (which basically corresponds to the number of line in the input file) - le me know if this will help in the debugging of issues (
also if log level will be setup to DEBUG it will now log predict_kwargs)

Comment (RafalSkolasinski) [reply python/seldon_core/batch_processor.py:413]: Also, now in the output file there will be single line of output for each line of input like with BATCH_SIZE=1 option

File: python/seldon_core/batch_processor.py
Comment (RafalSkolasinski) [reply python/seldon_core/batch_processor.py]: This should be documented now

File: python/seldon_core/batch_processor.py
Comment (RafalSkolasinski) [reply python/seldon_core/batch_processor.py]: This should be documented now

File: python/seldon_core/batch_processor.py
Comment (axsaucedo) [inline python/seldon_core/batch_processor.py:442]: You normally can just use logger.exception, as that prints the stack trace but this should be ok as well

File: python/seldon_core/batch_processor.py
Comment (axsaucedo) [inline python/seldon_core/batch_processor.py:379]: indexes is quite commonly used but we can change to indices

File: python/seldon_core/batch_processor.py
Comment (axsaucedo) [inline python/seldon_core/batch_processor.py]: Probably should make sure is typed (for consistency)

File: doc/source/servers/batch.md
Comment (axsaucedo) [inline doc/source/servers/batch.md:145]: may be worth adding a link to "micro batching" concept that explains what it is (if relevant)

File: doc/source/servers/batch.md
Comment (axsaucedo) [inline doc/source/servers/batch.md:186]: May be worth adding a reference to the issue where this is explained in more detail around corner cases

**Review by axsaucedo — APPROVED** [review]
NIce one - comments are quite minor and some are more suggestions so we can resolve if not relevant, and we can merge once tests pass
/approve

File: python/seldon_core/batch_processor.py
Comment (RafalSkolasinski) [reply python/seldon_core/batch_processor.py]: done

File: python/seldon_core/batch_processor.py
Comment (RafalSkolasinski) [reply python/seldon_core/batch_processor.py]: changed back

File: doc/source/servers/batch.md
Comment (RafalSkolasinski) [reply doc/source/servers/batch.md:145]: not sure what link you have in mind?

**Comment by seldondev** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *[axsaucedo](https://github.com/SeldonIO/seldon-core/pull/3718#pullrequestreview-798135652 "Approved")*

The full list of commands accepted by this bot can be found [here](https://jenkins-x.io/v3/develop/reference/chatops/?repo=SeldonIO%2Fseldon-core).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/SeldonIO/seldon-core/blob/master/OWNERS)~~ [axsaucedo]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: doc/source/servers/batch.md
Comment (RafalSkolasinski) [reply doc/source/servers/batch.md:186]: Done

File: python/seldon_core/batch_processor.py
Comment (yaliqin) [reply python/seldon_core/batch_processor.py:416]: Thanks for the modification. We currently use the batch processor in the pipeline. During the pipeline running I read this step log file for debugging. The line number is useful for debugging.

File: doc/source/servers/batch.md
Comment (axsaucedo) [reply doc/source/servers/batch.md:145]: I can't really find a resourcee that isn't a personal blog that explains this so probably we can just leave it

**Comment by seldondev** [comment]
@RafalSkolasinski: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integration | da7fc53a2fccf932266176c579061a6608dde27d | [link](http://dashboard-jx.34.66.17.27.nip.io/namespaces/jx/pipelineruns/seldon-core-pr-3718-integration-g4klj) | `/test integration`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by seldondev** [comment]
@RafalSkolasinski: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integration | da7fc53a2fccf932266176c579061a6608dde27d | [link](http://dashboard-jx.34.66.17.27.nip.io/namespaces/jx/pipelineruns/seldon-core-pr-3718-integration-sks95) | `/test integration`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by seldondev** [comment]
@RafalSkolasinski: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integration | da7fc53a2fccf932266176c579061a6608dde27d | [link](http://dashboard-jx.34.66.17.27.nip.io/namespaces/jx/pipelineruns/seldon-core-pr-3718-integration-nsvdv) | `/test integration`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by RafalSkolasinski** [comment]
Integration failure on: `TestPrepack.test_sklearn_v2` so unrelated to the PR

**Comment by RafalSkolasinski** [comment]
The executor failure 
```
--- FAIL: TestTimeout (1.00s)
    client_test.go:315: Started
    client_test.go:341: 
        Expected
            <string>: Get "http://127.0.0.1:35239/health/status": context deadline exceeded
        to contain substring
            <string>: Client.Timeout exceeded while awaiting headers
```
looks odd but if anything this is unrelated to this PR.

**Comment by seldondev** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *[axsaucedo](https://github.com/SeldonIO/seldon-core/pull/3718#issuecomment-961757679 "Approved")*

The full list of commands accepted by this bot can be found [here](https://jenkins-x.io/v3/develop/reference/chatops/?repo=SeldonIO%2Fseldon-core).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/SeldonIO/seldon-core/blob/master/OWNERS)~~ [axsaucedo]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Pull request merged**
